### PR TITLE
feat(api) : Accepting all types of Attribute values

### PR DESCRIPTION
**Pull request by rashidsp**
- Activate's check is covered in get_variation, however i think i should be separated to raise exception.

### CODE REVIEW

File: lib/optimizely/event_builder.rb [hunk 1]
```diff
@@ -15,10 +15,11 @@
...
 #
-require_relative './audience'
-require_relative './params'
-require_relative './version'
-require_relative '../optimizely/helpers/event_tag_utils'
+require_relative 'audience'
+require_relative 'helpers/constants'
+require_relative 'helpers/event_tag_utils'
+require_relative 'params'
+require_relative 'version'
```
Comment (oakbani) [inline lib/optimizely/event_builder.rb]: nit empty line before secure_random

File: lib/optimizely/event_builder.rb [hunk 2]
```diff
@@ -74,9 +75,9 @@ def get_common_params(user_id, attributes)
       visitor_attributes = []
 
       attributes&.keys&.each do |attribute_key|
-        # Omit null attribute values
+        # Omit null attribute invalid values
```
Comment (oakbani) [inline lib/optimizely/event_builder.rb]: What does #omit null attribute invalid means? 🤐 
Omit invalid attribute values. 

Does this also check that attribute key is a string?

File: lib/optimizely/helpers/constants.rb [hunk 3]
```diff
@@ -327,6 +327,8 @@ module Constants
         'v3' => '3',
         'v4' => '4'
       }.freeze
+
+      ATTRIBUTE_VALID_TYPES = [FalseClass, Fixnum, Float, TrueClass, String].freeze
```
Comment (oakbani) [inline lib/optimizely/helpers/constants.rb]: What about Integer?
And we need to check strictly False primitive type. 
We don't want to allow any class inheriting from False or True classes. Please make sure

File: lib/optimizely/helpers/validator.rb [hunk 4]
```diff
@@ -30,8 +30,17 @@ def attributes_valid?(attributes)
         # attributes - User attributes to be validated.
         #
         # Returns boolean depending on validity of attributes.
+        return false unless attributes.is_a?(Hash)
```
Comment (oakbani) [inline lib/optimizely/helpers/validator.rb]: Why changed? Previous was more precise.

File: lib/optimizely/helpers/validator.rb [hunk 5]
```diff
@@ -30,8 +30,17 @@ def attributes_valid?(attributes)
...
+      end
 
-        attributes.is_a?(Hash)
+      def attribute_value_type_valid?(attribute_value)
+        # Determines if provided attribute value is valid.
+        #
+        # attribute_value - value to be validated.
+        #
+        # Returns boolean depending on validity of attribute value.
+        !attribute_value.nil? && Helpers::Constants::ATTRIBUTE_VALID_TYPES.include?(attribute_value.class)
```
Comment (oakbani) [inline lib/optimizely/helpers/validator.rb]: I think it will do without !attribute_value.nil?

File: spec/project_spec.rb [hunk 6]
```diff
@@ -780,6 +782,12 @@ class InvalidErrorHandler; end
       project_instance.is_feature_enabled('multi_variate_feature', 'test_user')
     end
 
+    it 'should log and raise an exception when called with attributes in an invalid format' do
+      expect { project_instance.is_feature_enabled('multi_variate_feature', 'test_user', 'invalid_attributes') }
```
Comment (oakbani) [inline spec/project_spec.rb]: also assert that isFeatureEnabled returns False.

File: spec/project_spec.rb [hunk 7]
```diff
@@ -895,6 +903,12 @@ class InvalidErrorHandler; end
       project_instance.get_enabled_features('test_user', 'browser_type' => 'chrome')
     end
 
+    it 'should log and raise an exception when called with attributes in an invalid format' do
+      expect { project_instance.get_enabled_features('test_user', 'invalid_attributes') }
+        .to raise_error(Optimizely::InvalidAttributeFormatError)
```
Comment (oakbani) [inline spec/project_spec.rb]: also assert that it returns empty array

File: spec/project_spec.rb [hunk 8]
```diff
@@ -1211,6 +1225,12 @@ class InvalidErrorHandler; end
       expect(project_instance.get_feature_variable_integer('integer_single_variable_feature', 'integer_variable', nil, user_attributes))
         .to eq(nil)
     end
+
+    it 'should log and raise an exception when called with attributes in an invalid format' do
+      expect { project_instance.get_feature_variable_integer('integer_single_variable_feature', 'integer_variable', user_id, 'invalid_attributes') }
```
Comment (oakbani) [inline spec/project_spec.rb]: also assert that it returns None

File: lib/optimizely/event_builder.rb (hunk 2 above)
Comment (oakbani) [reply lib/optimizely/event_builder.rb]: @rashidsp

**Comment by rashidsp** [comment]
closing this. PR created at Optly

**Pull request closed without merging**